src/client/layout/layout_model3d.cpp

```diff
@@ -0,0 +1,236 @@
+    }
+    if (adoptedStyle.hasProperty("height"))
+    {
+      auto height = adoptedStyle.height();
+      if (!height.isAuto() &&
+          lastFragment.contentHeight() > 0)
+        adoptedHeight = height.toLayoutValue();
+    }
+
+    // If both width and height are specified, then use them.
+    if (adoptedWidth.has_value() && adoptedHeight.has_value())
+      return false;
+
+    // If both width and height are auto, then use default model size
+    if (!adoptedWidth.has_value() && !adoptedHeight.has_value())
+    {
+      // Use default size for 3D models (e.g., 200x200)
+      formattingContext().setContentSize(200.0f, 200.0f);
+      return true;
+    }
+
+    // If only one dimension is specified, use a square aspect ratio
+    if (adoptedWidth.has_value() && !adoptedHeight.has_value())
+    {
+      formattingContext().setContentSize(lastFragment.contentWidth(), lastFragment.contentWidth());
+      return true;
+    }
+    else if (!adoptedWidth.has_value() && adoptedHeight.has_value())
+    {
+      formattingContext().setContentSize(lastFragment.contentHeight(), lastFragment.contentHeight());
+      return true;
+    }
+
+    assert(false && "Unreachable");
+    return false;
+  }
+
+  void LayoutModel3d::setModelData(const std::vector<dom::HTMLModelElement::GaussianSplat> &splats)
+  {
+    adjustModelSize();
+
+    auto setModelData = [this, &splats](Scene &scene)
+    {
+      Model3d &modelComponent = scene.getComponentChecked<Model3d>(entity());
+
+      // Convert HTMLModelElement::GaussianSplat to builtin_scene::GaussianSplat
+      std::vector<builtin_scene::GaussianSplat> convertedSplats;
+      convertedSplats.reserve(splats.size());
+
+      for (const auto &splat : splats)
+      {
+        builtin_scene::GaussianSplat convertedSplat;
+        convertedSplat.position[0] = splat.position[0];
+        convertedSplat.position[1] = splat.position[1];
+        convertedSplat.position[2] = splat.position[2];
+        convertedSplat.color[0] = splat.color[0];
+        convertedSplat.color[1] = splat.color[1];
+        convertedSplat.color[2] = splat.color[2];
+        convertedSplat.opacity = splat.opacity;
+        convertedSplat.scale[0] = splat.scale[0];
+        convertedSplat.scale[1] = splat.scale[1];
+        convertedSplat.scale[2] = splat.scale[2];
+        convertedSplat.rotation[0] = splat.rotation[0];
+        convertedSplat.rotation[1] = splat.rotation[1];
+        convertedSplat.rotation[2] = splat.rotation[2];
+        convertedSplat.rotation[3] = splat.rotation[3];
+        convertedSplats.push_back(convertedSplat);
+      }
+
+      modelComponent.setSplats(std::move(convertedSplats));
+
+      // Update spatial information from the HTML element
+      assert(dom::Node::Is<dom::HTMLModelElement>(node()));
+      auto &modelElement = dom::Node::AsChecked<dom::HTMLModelElement>(node());
+
+      WebContent &webContent = scene.getComponentChecked<WebContent>(entity());
```

Comment:
The model element should `enableCustomGeometry()` and has no WebContent component.